[ffmpeg] Add feature support for zlib, iconv, fdk-aac, mp3lame, opus, soxr, theora.

Sibras opened the pull request:
* [zlib] Fix unistd.h being incorrectly required when imported from a project defining HAVE_UNISTD_H=0

* [ffmpeg] Fix zlib feature. Zlib was not currently being used despite being marked as a dependency. Zlib was quietly failing due to a missing unistd.h. An update was also required to detect the correct debug lib.

* [ffmpeg] Add additional features to enable:
    * iconv
    * fdk-aac
    * mp3lame
    * opus
    * soxr
    * theora
    * x265
    * vorbis
    * speex
    * snappy
    * wavpack
    * sdl2 (required for ffplay)

* [ffmpeg] Fix various issues with FindFFmpeg.cmake:
    * correctly detect debug and release libs
    * add missing postproc lib
    * add optional feature dependency detection

* [ogg] Enable output of pkg-config file (required for detection by ffmpeg and vorbis)

* [speex] Enable output of pkg-config file

* [ffmpeg] Set pkg-config to search for static configurations when building in static configuration

* [ffmpeg] Fixes for detection of opencl, soxr, lame, iconv, fdk, x264, x265, openssl, wavpack on linux/static builds

* [sdl2] Enable output of pkg-config file

* [fdk-aac] Correct header installation to "include/fdk-aac" to match fdk's normal build scripts (required for any lib such as ffmpeg to get the correct header location)

* [ffmpeg] Remove nvcodec's dependency on cuda

* [pkgconfig] Add additional common system libraries.

* [vcpkg_fixup_pkgconfig] Add to ffnvcodec, ogg, sdl2, speex, x265, wavpack, vorbis, opus

Fixes #7084, Fixes #9671, Fixes #7544

Sibras committed:
[zlib] Fix unistd.h being required

Sibras committed:
[ffmpeg] Fix zlib.

Closes #7084

Sibras committed:
[ffmpeg] Add iconv, fdk-aac, mp3lame, opus, soxr, theora support.

Closes #9671

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/vcpkg?pullRequest=11277) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/vcpkg?pullRequest=11277) before we can accept your contribution.<br/><br/>:x: Sibras [sign now](https://cla.opensource.microsoft.com/microsoft/vcpkg?pullRequest=11277)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/vcpkg?pullRequest=11277) it.</sub>

cenit commented:
A small subset of features is already under revision in #11151

Sibras commented:
I have seen #11151 but this is a superset of that which doesnt directly apply over the top of it without having to make some changes to #11151 (although minor ones). So since i had already done the work for lame and opus (the only 2 added by #11151) I left them in this patch as it makes it easier to add and test the others without dealing with conflicts. 

As to which is used, it doesn't bother me. I just had this patch ready to go, all checks passing, so i thought id put it up.

Sibras committed:
[libogg] Enable creation of pkg-config file

Sibras committed:
[ffmpeg] Add vorbis feature

cenit commented:
Thank you for your work!

Sibras committed:
[speex] Enable creation of pkg-config file

Sibras committed:
[ffmpeg] Add speex feature

Sibras committed:
[ffmpeg] Add snappy feature

Sibras committed:
[ffmpeg] Add wavpack feature

Sibras commented:
I also have commits that add support for vorbis, speex, snappy and wavpack features but ill wait for a review on these before pushing those

JackBoosY commented on ports/ffmpeg/CONTROL:
```diff
@@ -1,6 +1,5 @@
 Source: ffmpeg
-Version: 4.2-8
-Build-Depends: zlib
```
Should we set zlib as a default feature?

Sibras replied:
Previously it was listed as a default feature but ffmpegs configure was silently failing and never actually using it. So not having it included as default is actually consistent with the builds that vcpkg was previously generating

Sibras replied:
By default (non vcpkg builds) ffmpeg auto enables bzlib2, iconv, lzma, sdl2, zlib libraries. So if we were to auto enable libraries such as zlib then those others would also be on the list if we are trying to be consistent with that.

Sibras replied:
OK to fully cover the whole zlib depends issue and hopefully help others understand and avoid confusion in the long term here is a little FYI:

vcpkg's cmake file just sets some build options and then calls ffmpegs built in configure/make based build script. By default, unless told to do otherwise, ffmpegs build script will auto test for the availability and **usability** of some (not all) of its dependencies (zlib, bzip2, iconv, lzma, sdl2 and some other platform dependent options) and enable them. 

When building, if zlib was installed prior to ffmpeg then ffmpegs build script will detect its available and try and use it. But on windows builds due to the unistd.h error that this PR fixes it would be detected as **not usable** and would therefore not be enabled.

My guess is that the reason that zlib was listed as a build dependency was that someone built ffmpeg on non-windows using static linking (dynamic linking works fine). In this case there is a valid unistd.h and as such zlib was detected as **usable** by ffmpegs build script and automatically included. This would cause unforeseen link errors about missing zlib and give the incorrect impression that zlib was a required dependency. However this could also happen for any of the other dependencies that ffmpeg may auto enable (listed above) but only the zlib case was previously handled meaning all other cases would fail as well.

With this patch, vcpkgs cmake script is updated so that it now takes direct control over which external dependencies are used and prevents ffmpegs build script from attempting to auto enable any of them by passing it the appropriate configure options. As such there are now no hidden auto enabled dependencies and the only dependencies used are the ones vcpkg users explicitly enable.

JackBoosY commented on ports/zlib/0001-Prevent-invalid-inclusions-when-HAVE_-is-set-to-0.patch:
```diff
@@ -0,0 +1,53 @@
+ #endif
+ 
+ #ifdef HAVE_UNISTD_H    /* may be set to #if 1 by ./configure */
+-#  define Z_HAVE_UNISTD_H
```
Will zlib officially accept these changes?

Sibras replied:
No idea, its a fix that seems rather specific to ffmpeg on windows (im not aware of any other situation that it occurs). Someone is welcome to try it and find out.

Sibras replied:
A PR has been sent to zlib for official inclusion. Due to the slow pace of zlib releases it may take a while before it reaches a release.

cenit commented:
@Sibras when consuming ffmpeg, using cmake, is it necessary to modify FindFFMPEG.cmake in order to add missing dependencies, in case ffmpeg is built with some of your new features?
If you cannot try it on some OS, it’s enough to just put all features as default features *temporarily*, and see if CI is complaining. Maybe also try to turn  ffmpeg as a default feature also in opencv, since that is a known port dependent on a proper FindFFMPEG.cmake

Sibras committed:
[ffmpeg] Fix linking to static dependencies using pkg-config

Sibras committed:
[ffmpeg] Fix opencl detection on linux

Sibras committed:
[ffmpeg] Fix soxr linking on linux

Sibras committed:
[ffmpeg] Fix zlib detection on linux

Sibras commented:
@cenit Most likely FindFFMPEG.cmake should be modified, However many of the existing features were already not added to FindFFMPEG.cmake (such as openssl, x264 etc.) before this so it was already broken. Ill have a look at fixing though

cenit commented:
It’s not broken for default features for sure. I’d not be happy about merging even more features without support inside FindFFMPEG.cmake, but that’s just my feeling.

Sibras commented:
@cenit, OK ive looked over FindFFMPEG.cmake and it looks like it is actually broken even for defaults. It was incorrectly finding the wrong ffmpeg libs (only ever found debug) and was not searching for postproc.lib which is actually used by avfilter so you would still get link errors with static builds (linux and windows).

Ive fixed the above issues and added in handling for the previously missing dependencies + all the ones i have added. Ill update the pr later once i clean up the patches.

For reference just testing opencv is not enough as that just builds a lib. So for testing i created a 5 line program that pulled in ffmpeg using cmake and then tested it using linux+windows static build triplets. Not sure if something like that should/can be added to the CI to help test in the future.

cenit commented:
@sibras thanks for adding them. This kind of testing, while very important, is still out of scope for vcpkg. 
To be honest, I also have it. And I see ffmpeg detecting debug and release libs and correctly linking it towards a very small simple program, but of course I can just be in a working corner case. 
I hope to see your fixes soon :)

Sibras committed:
[ffmpeg] Fix FindFFmpeg.cmake

* Adds postproc to list of searched libs
* Adds missing windows dependency libs
* Fixes detection of seperate debug/release libs

Sibras committed:
[ffmpeg] Add optional dependencies to FindFFmpeg.cmake

Sibras committed:
[fdkaac] Correct header installation directory

This matches the installation directory used by the default fdkaac build scripts and is required for other existing projects attempting to find it.

Sibras committed:
[sdl2] Enable creation of pkg-config file on windows

Sibras committed:
[ffmpeg] Add sdl2 feature

Sibras committed:
[ffmpeg] Fix detection of static mp3lame

Sibras committed:
[ffmpeg] Fix detection of static iconv

Sibras commented:
bump

JackBoosY commented:
/azp run

JackBoosY commented on ports/ffmpeg/0006-fix-StaticFeatures.patch:
```diff
@@ -2,19 +2,48 @@ diff --git a/configure b/configure
  enabled opencl            && { check_pkg_config opencl OpenCL CL/cl.h clEnqueueNDRangeKernel ||
                                 check_lib opencl OpenCL/cl.h clEnqueueNDRangeKernel -Wl,-framework,OpenCL ||
                                 check_lib opencl CL/cl.h clEnqueueNDRangeKernel -lOpenCL ||
 +                               check_lib opencl CL/cl.h clEnqueueNDRangeKernel -lOpenCL -lAdvapi32 -lOle32 -lCfgmgr32||
```
Advapi32 only support windows.

Sibras replied:
Yes, so in the non-window case this check_lib call will fail as those libs dont exist. This is intended as the check_lib call is used to detect if the lib is available. So on windows it will succeed and then add the required libs to the final build. However, on non-windows it will silently fail and do nothing. This is intended behavior as one of the existing other check_lib calls will succeed on linux and add the required libs for that case.

JackBoosY commented:
Please resolve the file conflicts.

Thanks.

PhoebeHui commented:
/azp run

PhoebeHui commented:
@Sibras, thanks for the PR!

Have you test the new features locally?

Sibras commented:
@PhoebeHui Yes all features have been tested locally. I tested linux (in WSL) and windows builds in both shared and static.

Although a change was made after this PR was initially submitted that has broken x264 (see #11765). But that issue is unrelated to any of the changes in this PR.

Sibras commented:
force pushed just to resolve merge conflicts, no actual changes were made.

Sibras committed:
[ffmpeg] Use newer libvpx package detection.

JackBoosY commented:
/azp run

NancyLi1013 commented:
@Sibras 
Could you please resolve the conflicts?

JackBoosY commented:
@Sibras Could you resolve these file conflicts?

Thanks.

Sibras committed:
[ffmpeg] Add x265 support.

Sibras commented:
@NancyLi1013 @JackBoosY Done

JackBoosY commented:
@ras0219 @ras0219-msft Ping for response.
I think this issue is ready to merge now.

RamadanAhmed commented:
Hi,
Will this pull request be merged any time soon ?

ras0219-msft commented on ports/libogg/portfile.cmake:
```diff
@@ -11,7 +11,7 @@ vcpkg_from_github(
     SOURCE_PATH ${SOURCE_PATH}
     PREFER_NINJA
-    OPTIONS -DINSTALL_DOCS=0 -DINSTALL_PKG_CONFIG_MODULE=0
+    OPTIONS -DINSTALL_DOCS=0 -DINSTALL_PKG_CONFIG_MODULE=1
```
Does this need `vcpkg_fixup_pkgconfig()`?

https://github.com/microsoft/vcpkg/blob/master/scripts/cmake/vcpkg_fixup_pkgconfig.cmake

Sibras replied:
When I first created this PR the use of vcpkg_fixup_pkgconfig broke libogg (same for the other ports mentioned below). I havnt tested recently as getting this PR merged was taking a while as it was, so I was waiting for this before adding any new commits.

Sibras replied:
OK so vcpkg_fixup_pkgconfig is still broken for ffmpeg.
It converts occurrences of (for example)
`prefix=D:/Code_Files/vcpkg/packages/libogg_x64-windows`
to
`prefix=${prefix}`

which result in following errors
```
pkg-config --exists --print-errors libogg
Variable 'prefix' not defined in '/d/Code_Files/vcpkg/installed/x64-windows/lib/pkgconfig/libogg.pc'
```

The ffmpeg configure script is called with '--prefix' option set but pkg-config doesnt seem to know anything about it.

Sibras replied:
Ive added a patch to ffmpeg to manually pass 'prefix' variable to pkg-config. Im not sure if the output from vcpkg_fixup_pkgconfig() is the correct one longterm but passing prefix to pkg-config is the only way to fix it for ffmpeg. So it at least works now.

Sibras replied:
Ports now all use vcpkg_fixup_pkgconfig. All ports have been corrected to work with the latest version of vcpkg_fixup_pkgconfig.

ras0219-msft commented on ports/sdl2/portfile.cmake:
```diff
@@ -15,6 +15,7 @@ vcpkg_extract_source_archive_ex(
         fix-space-in-path.patch
         disable-wcslcpy-and-wcslcat-for-windows.patch
         fix-EventToken-header-reference.patch
+        0006-sdl2-Enable-creation-of-pkg-cfg-file-on-windows.patch
```
Does this need `vcpkg_fixup_pkgconfig()`?

https://github.com/microsoft/vcpkg/blob/master/scripts/cmake/vcpkg_fixup_pkgconfig.cmake

Sibras replied:
Ive added vcpkg_fixup_pkgconfig() to this and all other ports currently used by ffmpeg.

ras0219-msft commented on ports/speex/portfile.cmake:
```diff
@@ -6,6 +6,8 @@ vcpkg_from_github(
   SHA512  612dfd67a9089f929b7f2a613ed3a1d2fda3d3ec0a4adafe27e2c1f4542de1870b42b8042f0dcb16d52e08313d686cc35b76940776419c775417f5bad18b448f
   HEAD_REF master
+  PATCHES
+    0001-make-pkg-config-lib-name-configurable.patch
```
Does this need `vcpkg_fixup_pkgconfig()`?

https://github.com/microsoft/vcpkg/blob/master/scripts/cmake/vcpkg_fixup_pkgconfig.cmake

Sibras replied:
See above (Ive added vcpkg_fixup_pkgconfig() to this and all other ports currently used by ffmpeg.)

ras0219-msft reviewed (commented):
Overall LGTM, I had a few questions that might not require changes to resolve.

Ready to merge once the above questions are addressed!

Sibras commented on ports/ffmpeg/CONTROL:
```diff
@@ -13,51 +12,104 @@ Feature: ffserver
 Description: ffserver appplication support in ffmpeg
 
 Feature: ffplay
+Build-Depends: sdl2
```
sdl2 when enabled is used by ffmpeg as an optional feature but its a required feature for ffplay. So your right this should be ffmpeg[sdl2]. Changed the PR accordingly.

Sibras committed:
[ffmpeg] Use newer zlib pkg-config detection.

Sibras commented on ports/ffmpeg/0004-fix-debug-build.patch:
```diff
@@ -12,19 +12,39 @@ index bd2de34..fba948a 100755
 -enabled bzlib && check_lib bzlib bzlib.h BZ2_bzlibVersion    -lbz2
 -enabled  lzma && check_lib lzma   lzma.h lzma_version_number -llzma
 +if enabled debug_configure; then
++    enabled  zlib && { check_pkg_config zlib zlib "zlib.h" zlibVersion ||
```
It does, Ive updated the PR accordingly.

Sibras committed:
[ffmpeg] Fix detection of 'prefix' in pkgconfig

Required to get current output from vcpkg_fixup_pkgconfig to work with ffmpegs configure script.

Sibras committed:
[ffnvcodec] Use vcpkg_fixup_pkgconfig.

Sibras committed:
[libogg] Use vcpkg_fixup_pkgconfig.

Sibras committed:
[sdl2] Correct pkgconfig debug suffix.

Sibras committed:
[speex] Use vcpkg_fixup_pkgconfig.

Sibras committed:
[x265] Fix pkgconfig windows static lib naming.

Sibras committed:
[wavpack] Fix pkgconfig windows dll naming.

JackBoosY commented:
Please resolve these regressions, you can get the failure logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=40413&view=artifacts&type=publishedArtifacts).

Sibras committed:
[ffnvcodec] Install pkgconfig file in debug.

Sibras committed:
[libvorbis] Fixup pkgconfig file.

Sibras committed:
[opus] Use vcpkg_fixup_pkgconfig.

Sibras committed:
[ffmpeg] Correctly set debug lib folder with pkgconfig.

Sibras commented:
> Please resolve these regressions, you can get the failure logs here.

All are due to vcpkg_fixup_pkgconfig. Ive added options that fix most of the errors bit it looks like #11050 is needed to resolve all the issues.

JackBoosY commented:
```
-- Checking file: /mnt/vcpkg-ci/packages/sdl2_x64-linux/lib/pkgconfig/sdl2.pc
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:150 (message):
  Found -Wl,-rpath,${libdir} and no rule to analyse the flag! Please check
  the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:234 (vcpkg_fixup_pkgconfig_check_libraries)
  ports/sdl2/portfile.cmake:90 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:79 (include)


Error: Building package sdl2:x64-linux failed with: BUILD_FAILED
```

Sibras commented:
As mentioned previously, there is a bug in current vcpkg_fixup_pkgconfig that causes the above error. Its a simple 1 word fix but given that there is #11550 that also fixes the issue I mentioned that this should wait for that. I can fix it in this PR if desired.

Sibras committed:
[vcpkg/scripts/pkgconfig] Add additional common definitions

Sibras commented:
Force pushed just to resolve merge conflicts, only new changes are related to issues with current vcpkg_fixup_pkgconfig.

ras0219-msft commented:
/azp run

endrift commented:
While you're here, can you add libwebp too?

Sibras committed:
Merge branch 'master' into master

Sibras commented:
> 
> 
> While you're here, can you add libwebp too?

Ive got a few extra changes/additions that i was planning on making but Im waiting to get this merged first.

Sibras committed:
[sdl2] Fix pkgconfig on osx.

JackBoosY commented:
pbc regression will be fixed in #12732

PhoebeHui commented:
Features test with x64-window:
Except 'x264', other features test pass with x64-windows, x264 failed on windows due to issue https://github.com/microsoft/vcpkg/issues/11765 to track.

Features test with linux: in progress

JackBoosY commented:
The sdl2 regression is still exist:
```
-- Checking file: /Users/vagrant/Data/packages/sdl2_x64-osx/lib/pkgconfig/sdl2.pc
-- CHECK_LIB_iconv_RELEASE:CHECK_LIB_iconv_RELEASE-NOTFOUND
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:239 (message):
  Library "iconv" was not found! If it is a system library use the
  SYSTEM_LIBRARIES parameter for the vcpkg_fixup_pkgconfig call! Otherwise,
  correct the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:316 (vcpkg_fixup_pkgconfig_check_files)
  ports/sdl2/portfile.cmake:97 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:79 (include)


Error: Building package sdl2:x64-osx failed with: BUILD_FAILED
```

PhoebeHui reviewed (changes requested):
the fdk-aac lib named 'libfdk-aac' on linux.

Sibras committed:
[ffmpeg] Correctly detect fdkac on linux.

Co-authored-by: Phoebe <20694052+PhoebeHui@users.noreply.github.com>

Sibras commented:
> -- Checking file: /Users/vagrant/Data/packages/sdl2_x64-osx/lib/pkgconfig/sdl2.pc
> -- CHECK_LIB_iconv_RELEASE:CHECK_LIB_iconv_RELEASE-NOTFOUND
> CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:239 (message):
>   Library "iconv" was not found! If it is a system library use the
>   SYSTEM_LIBRARIES parameter for the vcpkg_fixup_pkgconfig call! Otherwise,
>   correct the *.pc file
> Call Stack (most recent call first):
>   scripts/cmake/vcpkg_fixup_pkgconfig.cmake:316 (vcpkg_fixup_pkgconfig_check_files)
>   ports/sdl2/portfile.cmake:97 (vcpkg_fixup_pkgconfig)
>   scripts/ports.cmake:79 (include)
> 
> 
> Error: Building package sdl2:x64-osx failed with: BUILD_FAILED

I dont have osx in order to track this down. SDL only includes iconv if it is detected. So the fact that it is added to the pc file indicates that its cmake script found the library. But vcpkg_fixup_pkgconfig does not find it and I dont know why without access to an osx machine.

Sibras commented:
OK further investigation shows that the vcpkg iconv port does nothing on a non-windows system, so sdl must be detecting a system iconv.

Sibras committed:
[vcpkg/scripts/pkgconfig] Add iconv as system lib on osx.

PhoebeHui commented:
Features test with linux:
Except below features, other features test passed.

- [ ] avisynthplus: the dependency port failed
- [ ] nvcodec: depends on cuda
- [ ] speex: the dependency port failed
- [ ] x265: ERROR: x265 not found using pkg-config
- [ ] openssl: openssl not found
- [X] wavpack: the dependency port failed, fixed by https://github.com/microsoft/vcpkg/pull/12742
- [ ] fdk-aac, libfdk_aac not found, should be fixed with suggested changes.

Sibras committed:
[ffmpeg] Fix fdk detection on linux.

Sibras committed:
[ffmpeg] Fix x264 detection on linux.

Sibras committed:
[ffmpeg] Fix x265 detection on linux.

Sibras committed:
[ffmpeg] Fix detection of OpenSSL 1.1.0+ on linux.

Sibras committed:
[ffmpeg] Remove nvcodec dependency on cuda.

Sibras commented:
Only outstanding issues on linux are now:
* avisynthplus: this port only supports dynamic linkage and so does not support linux
* speex: Port currently does not support linux

JackBoosY commented on ports/ffmpeg/portfile.cmake:
```diff
@@ -93,58 +93,181 @@ else()
-if("avresample" IN_LIST FEATURES)
-    set(OPTIONS "${OPTIONS} --enable-avresample")
+if("nvcodec" IN_LIST FEATURES)
+    set(OPTIONS "${OPTIONS} --enable-cuda --enable-nvenc --enable-cuvid --disable-libnpp")
```
I noticed `libnpp` always be disabled, so please set `OPTIONS` to `--disable-libnpp` in line 64.

JackBoosY replied:
`cuda` is a feature, so please add a new feature `cuda` and set other features like `nvcodec` depend on it.

Sibras replied:
`    set(OPTIONS "${OPTIONS} --enable-cuda --enable-nvenc --enable-cuvid --disable-libnpp")`
was the previous settings prior to this PR (the patch just moved the line) which I didnt change. However it was not entirely correct. The "cuda" feature is an old leftover and that line doesnt really do much, it does not mean that ffmpeg depends on cuda (its not well named). Its just there to ensure correct feature selection. The ffnvcodec feature does not depend on cuda nor did it ever, that dependency was incorrectly set prior to this PR. The libnpp feature is an entirely different one that shouldnt have been on that line either. I left it as it was but ill now instead add a new commit that removes it.

Sibras replied:
The 'cuda' command line option is just not well named as it doesnt mean a dependency on cuda. For reference the feature that does depend on cuda is called 'cuda-sdk' and 'cuda-nvcc' which are entirely different features and need to be added in a new PR.

JackBoosY replied:
Could you leave comments to this line?

Sibras replied:
I was going to make a separate PR later that added the actual cuda-sdk features (this PR is busy enough as it is). Did you want me to add a comment then or to this PR. Also what specifically did you want commented on?

JackBoosY replied:
I think you'd better to leave comment here.

Sibras replied:
Added extra commit adding a comment to portfile

PhoebeHui commented on ports/ffmpeg/FindFFMPEG.cmake.in:
```diff
@@ -47,61 +43,182 @@ if(@ENABLE_BZIP2@)
+
+if(@ENABLE_WAVPACK@)
+  find_dependency(wavpack)
+  get_target_property(WAVPACK_LIBRARY_RELEASE Wavpack::wavpack IMPORTED_LOCATION_RELEASE)
+  get_target_property(WAVPACK_LIBRARY_DEBUG Wavpack::wavpack IMPORTED_LOCATION_DEBUG)
```
```suggestion
  get_target_property(WAVPACK_LIBRARY_RELEASE WavPack::wavpack IMPORTED_LOCATION_RELEASE)
  get_target_property(WAVPACK_LIBRARY_DEBUG WavPack::wavpack IMPORTED_LOCATION_DEBUG)
```

Sibras committed:
[ffmpeg] Use WavPack instead of Wavpack in cmake.

Co-authored-by: Phoebe <20694052+PhoebeHui@users.noreply.github.com>

Sibras committed:
[ffmpeg] Remove libnpp feature from nvcodec.

This feature is entirely unrelated and should be added in a new PR.

Sibras commented on scripts/cmake/vcpkg_common_definitions.cmake:
```diff
@@ -122,12 +122,61 @@ if(VCPKG_TARGET_IS_LINUX OR VCPKG_TARGET_IS_ANDROID OR VCPKG_TARGET_IS_OSX)
+
 # Windows system libs
 if(VCPKG_TARGET_IS_WINDOWS)
+    list(APPEND VCPKG_SYSTEM_LIBRARIES advapi32)
```
These are native windows libraries that are always available which is why I added them to VCPKG_SYSTEM_LIBRARIES. That variable is only used by vcpkg_fixup_pkgconfig so i dont see how it could cause any issues with other ports as its simply giving vcpkg_fixup_pkgconfig a more complete list of system libraries not managed by vcpkg. If im missing something could you explain how this could possibly cause any issues.

JackBoosY replied:
Since this macro only provides to `vcpkg_fixup_pkgconfig`, I can accept this.

Sibras committed:
Merge branch 'master' into master

Sibras committed:
[ffmpeg] Add comment about cuda options not requiring the cuda port as dependency.

PhoebeHui commented:
@Sibras, thanks for your hard work!

Test all features with latest changes:
x64-windows: pass
x64-linux: except the below 2 features, others passed.

x265: ERROR: libx265 not found
wavpack: ERROR: libwavpack not found

Sibras committed:
Merge remote-tracking branch 'upstream/master' into master

Sibras commented:
> 
> 
> @Sibras, thanks for your hard work!
> 
> Test all features with latest changes:
> x64-windows: pass
> x64-linux: except the below 2 features, others passed.
> 
> x265: ERROR: libx265 not found
> wavpack: ERROR: libwavpack not found
> 
> Double checked, the wavpack requires '-lm' library,
> 
> wavpack failure log:
> 
> ```
> /home/phoebe/11277/vcpkg/installed/x64-linux/debug/lib/../lib/libwavpack.a(pack_utils.c.o): In function `WavpackSetConfiguration64':
> /home/phoebe/11277/vcpkg/buildtrees/wavpack/src/0eba4c0c9f-ecbdfe4a39.clean/src/pack_utils.c:305: undefined reference to `floor'
> /home/phoebe/11277/vcpkg/buildtrees/wavpack/src/0eba4c0c9f-ecbdfe4a39.clean/src/pack_utils.c:311: undefined reference to `floor'
> /home/phoebe/11277/vcpkg/installed/x64-linux/debug/lib/../lib/libwavpack.a(pack.c.o): In function `pack_init':
> /home/phoebe/11277/vcpkg/buildtrees/wavpack/src/0eba4c0c9f-ecbdfe4a39.clean/src/pack.c:59: undefined reference to `floor'
> /home/phoebe/11277/vcpkg/installed/x64-linux/debug/lib/../lib/libwavpack.a(pack_dns.c.o): In function `dynamic_noise_shaping':
> /home/phoebe/11277/vcpkg/buildtrees/wavpack/src/0eba4c0c9f-ecbdfe4a39.clean/src/pack_dns.c:137: undefined reference to `floor'
> /home/phoebe/11277/vcpkg/buildtrees/wavpack/src/0eba4c0c9f-ecbdfe4a39.clean/src/pack_dns.c:140: undefined reference to `floor'
> /home/phoebe/11277/vcpkg/installed/x64-linux/debug/lib/../lib/libwavpack.a(pack_dns.c.o):/home/phoebe/11277/vcpkg/buildtrees/wavpack/src/0eba4c0c9f-ecbdfe4a39.clean/src/pack_dns.c:189: more undefined references to `floor' follow
> ```
> 
> x265 failure log
> 
> ```
> gcc -Wl,--as-needed -Wl,-z,noexecstack -pthread -o /tmp/ffconf.b3WzRGjY/test /tmp/ffconf.b3WzRGjY/test.o -lx265 -lm -ldl -lstdc++ -lgcc_s -lgcc -lrt
> /home/phoebe/11277/vcpkg/installed/x64-linux/lib/../lib/libx265.a(threadpool.cpp.o): In function `x265::ThreadPool::create(int, int, unsigned long)':
> threadpool.cpp:(.text+0x21c): undefined reference to `numa_available'
> threadpool.cpp:(.text+0x321): undefined reference to `numa_allocate_nodemask'
> /home/phoebe/11277/vcpkg/installed/x64-linux/lib/../lib/libx265.a(threadpool.cpp.o): In function `x265::ThreadPool::setThreadNodeAffinity(void*)':
> threadpool.cpp:(.text+0x585): undefined reference to `numa_available'
> threadpool.cpp:(.text+0x591): undefined reference to `numa_run_on_node_mask'
> threadpool.cpp:(.text+0x599): undefined reference to `numa_set_interleave_mask'
> /home/phoebe/11277/vcpkg/installed/x64-linux/lib/../lib/libx265.a(threadpool.cpp.o): In function `x265::ThreadPool::getNumaNodeCount()':
> threadpool.cpp:(.text+0x835): undefined reference to `numa_available'
> threadpool.cpp:(.text+0x843): undefined reference to `numa_max_node'
> /home/phoebe/11277/vcpkg/installed/x64-linux/lib/../lib/libx265.a(threadpool.cpp.o): In function `x265::ThreadPool::allocThreadPools(x265_param*, int&, bool)':
> threadpool.cpp:(.text+0x9bb): undefined reference to `numa_available'
> threadpool.cpp:(.text+0xc34): undefined reference to `numa_allocate_cpumask'
> threadpool.cpp:(.text+0xc5c): undefined reference to `numa_bitmask_weight'
> threadpool.cpp:(.text+0xc73): undefined reference to `numa_node_to_cpus'
> threadpool.cpp:(.text+0xca9): undefined reference to `numa_bitmask_free'
> threadpool.cpp:(.text+0x12b6): undefined reference to `numa_available'
> threadpool.cpp:(.text+0x131c): undefined reference to `numa_allocate_cpumask'
> threadpool.cpp:(.text+0x1330): undefined reference to `numa_bitmask_free'
> /home/phoebe/11277/vcpkg/installed/x64-linux/lib/../lib/libx265.a(threadpool.cpp.o): In function `x265::ThreadPool::~ThreadPool()':
> threadpool.cpp:(.text+0x566): undefined reference to `numa_bitmask_free'
> /home/phoebe/11277/vcpkg/installed/x64-linux/lib/../lib/libx265.a(threadpool.cpp.o): In function `x265::ThreadPool::setThreadNodeAffinity(void*)':
> threadpool.cpp:(.text+0x59f): undefined reference to `numa_set_localalloc'
> collect2: error: ld returned 1 exit status
> ERROR: libx265 not found
> ```

I cant reproduce either of these issues (Ubuntu 20.04). Both wavpack and x265 build fine (also seem to be working on the CI as well).

Sibras committed:
[ffmpeg] Fix static linking of wavpack on linux.

Sibras committed:
[ffmpeg] Fix x265 detection when using numa on linux.

Sibras commented:
> 
> @Sibras, thanks for your hard work!
> 
> Test all features with latest changes:
> x64-windows: pass
> x64-linux: except the below 2 features, others passed.
> 
> x265: ERROR: libx265 not found
> wavpack: ERROR: libwavpack not found
> 

Wavpack is fixed now. I couldnt reproduce the x265 error on my machine, I have libnuma-dev installed but its not picking it up. So i think I fixed the x265 error but I cant reproduce it to test.

PhoebeHui commented:
@Sibras, after I installed libnuma-dev, the feature 'x265' passed on linux. the feature 'wavpack' also test pass on linux.

Summary the feature tests status:
x64-linux: feature 'avisynthplus' and 'speex' doesn't support on linux. Others features test pass.
x64-windows: feature 'x264' failed, currently the issue in dependency port x264 has been fixed on master, however I suggest to fix the feature in sperate PR.

Currently no issues to block this PR, @strega-nil, could you please help review and merge this PR?

strega-nil commented:
@Sibras could you merge with master?

Sibras committed:
Merge branch 'master' into master

BillyONeal commented:
Thanks for your contribution!
